src/foam/nanos/auth/CachedUserAndGroupAuthService.java

```diff
@@ -0,0 +1,74 @@
+package foam.nanos.auth;
+
+import foam.core.X;
+import foam.nanos.util.LRULinkedHashMap;
+import java.util.Map;
+
+/**
+ * Created by marcroopchand on 2017-05-30.
+ */
+public class CachedUserAndGroupAuthService
+  extends UserAndGroupAuthService
+{
+  /**
+   * The cached data structure will look like this
+   *{
+   *  user1: {
+   *    permission1: true,
+   *    permission2: false
+   *    ...
+   *  },
+   *  user2: {
+   *    permission1: false,
+   *    permission2: true
+   *    ...
+   *  }
+   *  ...
+   *}
+   */
+  protected Map<String, Map<String, Boolean>> userMap = new LRULinkedHashMap<>(1000000);
+  protected Map<String, Boolean> permissionMap;
+
+  @Override
+  public void start() {
+    super.start();
+  }
+
+  @Override
+  public Boolean check(foam.core.X x, java.security.Permission permission) {
+    if ( x == null || permission == null ) return false;
+
+    User user = (User) x.get("user");
+    if ( user == null ) return false;
+
+    Group group = (Group) user.getGroup();
+    if ( group == null ) return false;
+
+    if ( userMap.containsKey(user.getId()) ) {
+      permissionMap = userMap.get(user.getId());
+    }
+    else {
+      permissionMap = new LRULinkedHashMap<>(1000);
```

Comment:
I think 1000 needs to be larger, or even just make this a regular HashMap. A user can easily be checked against thousands of permissions if you have one for each menu and one for each property and mode. The users will get cleaned up anyway, so this is redundant.